source/gui/messageDialog.py

```diff
@@ -0,0 +1,722 @@
+		:raises ValueError: If the action with the given id does not close the dialog.
+		:return: The updated dialog instance.
+		"""
+		if id not in (EscapeCode.DEFAULT, EscapeCode.NONE):
+			if id not in self._commands:
+				raise KeyError(f"No command registered for {id=}.")
+		if not self._commands[id].closes_dialog:
+			raise ValueError("Default actions that do not close the dialog are not supported.")
+		super().SetEscapeId(id)
+		return self
+
+	def setDefaultAction(self, id: ReturnCode | EscapeCode) -> Self:
+		"""See MessageDialog.SetEscapeId."""
+		return self.SetEscapeId(id)
+
+	def Show(self, show: bool = True) -> bool:
+		"""Show a non-blocking dialog.
+
+		Attach buttons with button handlers
+
+		:param show: If True, show the dialog. If False, hide it. Defaults to True.
+		"""
+		if not show:
+			return self.Hide()
+		if not self.GetMainButtonIds():
+			raise RuntimeError("MessageDialogs cannot be shown without buttons.")
+		self._realize_layout()
+		log.debug("Showing")
+		ret = super().Show(show)
+		if ret:
+			log.debug("Adding to instances")
+			self._instances.append(self)
+		return ret
+
+	def ShowModal(self) -> ReturnCode:
+		"""Show a blocking dialog.
+		Attach buttons with button handlers"""
+		if not self.GetMainButtonIds():
+			raise RuntimeError("MessageDialogs cannot be shown without buttons.")
+		self._realize_layout()
+		self.__ShowModal = self.ShowModal
+		self.ShowModal = super().ShowModal
+		from .message import displayDialogAsModal
+
+		log.debug("Adding to instances")
+		self._instances.append(self)
+		log.debug("Showing modal")
+		ret = displayDialogAsModal(self)
+		self.ShowModal = self.__ShowModal
+		return ret
+
+	@property
+	def isBlocking(self) -> bool:
+		"""Whether or not the dialog is blocking"""
+		return self.IsModal() and self.hasDefaultAction
+
+	@property
+	def hasDefaultAction(self) -> bool:
+		"""Whether the dialog has a valid default action."""
+		escapeId = self.GetEscapeId()
+		return escapeId != EscapeCode.NONE and (
+			any(command in (ReturnCode.CANCEL, ReturnCode.OK) for command in self._commands)
+			if escapeId == EscapeCode.DEFAULT
+			else True
+		)
+
+	# endregion
+
+	# region Public class methods
+	@classmethod
+	def CloseInstances(cls) -> None:
+		"""Close all dialogs with a default action"""
+		for instance in cls._instances:
+			if not instance.isBlocking:
+				instance.Close()
+
+	@classmethod
+	def BlockingInstancesExist(cls) -> bool:
+		"""Check if dialogs are open without a default return code
+		(eg Show without `self._defaultReturnCode`, or ShowModal without `wx.CANCEL`)"""
+		return any(dialog.isBlocking for dialog in cls._instances)
+
+	@classmethod
+	def FocusBlockingInstances(cls) -> None:
+		"""Raise and focus open dialogs without a default return code
+		(eg Show without `self._defaultReturnCode`, or ShowModal without `wx.CANCEL`)"""
+		for dialog in cls._instances:
+			if dialog.isBlocking:
+				dialog.Raise()
+				dialog.SetFocus()
+				break
```

Comment:
maybe this should happen for all instances? i.e. all get raised and maybe the last one gets focused?